src/Client/webpack.config.js

```diff
@@ -1,76 +1,116 @@
+            "@babel/polyfill",
+            resolve("./Client.fsproj")
+        ],
+    outputDir: resolve("./public"),
+    devServerPort: undefined,
+    devServerProxy: {
+        '/api/*': {
+            target: 'http://localhost:' + (process.env.SUAVE_FABLE_PORT || "8085"),
+            changeOrigin: true
+        }
+    },
+    historyApiFallback: {
+        index: resolve("./index.html")
+    },
+    contentBase: __dirname,
+    // Use babel-preset-env to generate JS compatible with most-used browsers.
+    // More info at https://github.com/babel/babel/blob/master/packages/babel-preset-env/README.md
+    babel: {
+        presets: [
+            ["@babel/preset-env", {
+                "targets": {
+                    "browsers": ["last 2 versions"]
+                },
+                "modules": false,
+                "useBuiltIns": "usage",
+            }]
+        ],
+        plugins: ["@babel/plugin-transform-runtime"]
+    }
+}
 
 var isProduction = process.argv.indexOf("-p") >= 0;
-var port = process.env.SUAVE_FABLE_PORT || "8085";
 console.log("Bundling for " + (isProduction ? "production" : "development") + "...");
 
+var path = require("path");
+var webpack = require("webpack");
+var MinifyPlugin = require("terser-webpack-plugin");
+
+var commonPlugins = [
+];
+
 module.exports = {
-  devtool: "source-map",
-  entry: resolve('./Client.fsproj'),
-  mode: isProduction ? "production" : "development",
-  output: {
-    path: resolve('./public'),
-    publicPath: "/public",
-    filename: "bundle.js"
-  },
-  resolve: {
-    symlinks: false,
-    modules: [ resolve("../../node_modules/")]
-  },
-  devServer: {
-    proxy: {
-      '/api/*': {
-        target: 'http://localhost:' + port,
-        changeOrigin: true
-      }
+    entry: CONFIG.fsharpEntry,
+    // NOTE we add a hash to the output file name in production
+    // to prevent browser caching if code changes
+    output: {
+        path: CONFIG.outputDir,
+        publicPath: "/public",
+        filename: '[name].js'
     },
-    hot: true,
-    inline: true,
-    historyApiFallback: true,
-  },
-  module: {
-    rules: [
-      {
-        test: /\.fs(x|proj)?$/,
-        use: {
-          loader: "fable-loader",
-          options: {
-            babel: babelOptions,
-            define: isProduction ? [] : ["DEBUG"]
-          }
-        }
-      },
-      {
-        test: /\.js$/,
-        exclude: /node_modules/,
-        use: {
-          loader: 'babel-loader',
-          options: babelOptions
+    resolve: {
+        symlinks: false,
+    },
+    mode: isProduction ? "production" : "development",
+    devtool: isProduction ? undefined : "source-map",
+    optimization: {
+        // Split the code coming from npm packages into a different file.
+        // 3rd party dependencies change less often, let the browser cache them.
+        splitChunks: {
+            cacheGroups: {
+                commons: {
+                    test: /node_modules/,
```

Comment:
What do you think about changing this to ``/node_modules|\.nuget/``?

I did a quick check with my own app:

|    file  | /node_modules/ | ``/node_modules\|\.nuget/`` |
| ---------- | ------ | ------ |
| main.js    | 157 kb | 30 kb  |
| vendors.js | 236 kb | 382 kb |
| total | 393 kb | 412 kb |

So the total size is ~5% greater, but the main bundle is much smaller.

Reply:
I am surprised that `.nuget` is the rule to apply here... Because in theory Fable 2 is using the `.fable` created locally.

Splitting between `vendors.js` and `main.js` is to optimize the browser caching. If you include `nuget` (aka Fable libraries in it) then if you use a new Fable feature or a new feature from a library then the `vendors.js` will change and will be loading from the server and not the cache.

Also, in general in the different projects, I use the hashed names but here because we are supporting SSR then I am not using the hashed names. We could support hashed names if we generate the `main.[hash].js` and `vendors.[hash].js` first and then update a file on the server to register the generated names.

To answer more directly your question, I am not against including more things in the `vendors.js` I think it's up to the people maintaining the SAFE stack to step in and discuss/decide.

Reply:
I'm still using Fable 1.x, so with fable 2.x it would probably be ``.fable`` as you say.


``vendors.js`` is still tree-shaken, right? So if I don't use a feature from a node lib, it would also get added / removed?

There is a manifest plugin for webpack - it will create a json file which maps from unhashed to hashed: https://github.com/webdeveric/webpack-assets-manifest

On the SSR side, this json file would need to be read.

Reply:
> `vendors.js` is still tree-shaken, right?

It should depending on the library. Also, we kind of consider that the `npm` usage change less often. Compare to Fable code but we can be wrong. Also, in general cache mechanism should be optimize depending the project setup/needs. It's hard to find a perfect rule for every one because each way have it's pros and cons. Also, I am not an expert of this things so can be wrong too :)

Indeed the maniafest plugin need to be added when using hashed name so the same content generate the same hash.